src/react/ISV/hooks/useIAMUser.tsx

```diff
@@ -0,0 +1,111 @@
+import { useMutation } from 'react-query';
+import { useMemo, useState, useEffect } from 'react';
+import { useIAMClient } from '../../IAMProvider';
+import { useAssumeRoleQuery } from '../../next-architecture/ui/S3ClientProvider';
+
+interface IAMUser {
+  id: string;
+  name: string;
+}
+
+type IAMUserStatus = 'loading' | 'success' | 'error';
+
+interface UseIAMUserProps {
+  IAMUserName: string;
+  IAMUserNameType: 'create' | 'existing';
+  onIAMUsersLoaded?: (users: IAMUser[]) => void;
+  onShouldGenerateKey?: (shouldGenerate: boolean) => void;
+}
+
+export const useIAMUser = ({
+  IAMUserName,
+  IAMUserNameType,
+  onIAMUsersLoaded,
+  onShouldGenerateKey,
+}: UseIAMUserProps) => {
+  const IAMClient = useIAMClient();
+  const { getQuery } = useAssumeRoleQuery();
+  const [status, setStatus] = useState<IAMUserStatus>('loading');
+  const [users, setUsers] = useState<IAMUser[]>([]);
+  const [accessKeys, setAccessKeys] = useState<string[] | null>(null);
+  const checkUserAccessKeys = async (userName: string) => {
+    try {
+      const userExists = users.some((user) => user.name === userName);
+      if (!userExists) {
+        return true;
+      }
+
+      const { AccessKeyMetadata } = await IAMClient.listAccessKeys(userName);
+      const activeKeys = AccessKeyMetadata.filter(
+        (key) => key.Status === 'Active',
+      );
+      if (activeKeys.length > 0) {
+        setAccessKeys(activeKeys.map((key) => key.AccessKeyId));
+      }
+      const shouldGenerateKey = !AccessKeyMetadata.some(
+        (key) => key.Status === 'Active',
+      );
+      onShouldGenerateKey?.(shouldGenerateKey);
+      return shouldGenerateKey;
+    } catch (error) {
+      onShouldGenerateKey?.(true);
+      return true;
+    }
+  };
+
+  useEffect(() => {
+    if (IAMUserNameType === 'existing' && IAMUserName) {
+      checkUserAccessKeys(IAMUserName);
+    }
+  }, [IAMUserName, IAMUserNameType]);
```

Comment:
This promise should be called within a useQuery hook in order to handle failure status, loading status and component unmount while the promise is ongoing safely.